autotools.py: add automatic depends on autotools when='@master,develop'

bernhardkaindl opened the pull request:
Practically every `AutotoolsPackage`, when built from git, will have to
run `autoreconf` to generate the `configure` script and other files.

Before calling `autoreconf`, `AutotoolsPackage.autoreconf()` checks
that `autoconf, automake` and `libtool` are installed,
as these are needed for `autoreconf` to succeed.

In the past, many `package.py` files were written without this check
present or working. As result many just have 'depends_on('libtool'),
bit miss it for `autoconf` and `automake`.

Also at the moment, that check is not working, as can be seen in #26005

1st:

Fix this check which didn't stop the code in #26005 by checking if the `autotools` packages are in `specs._dependencies`, not just in `specs`.

2nd:

With this fix, each package with version master or develop need to have these depends present to build for `@master` or `@develop`
```
    depends_on('autoconf', when='@master,develop')
    depends_on('automake', when='@master,develop')
    depends_on('libtool',  when='@master,develop')
```
In order to not have to update each relevant package.py, this commit
also makes adding these depends automatic for all AutotoolsPackages.

Of course it should excempt m4, autoconf, automake and libtool from
this to not add dependencies which they don't need to their graph,
a dependency on the package itself is also impossible.

To allow for this, the depends_on() gets an optional skip-argument for
which `AutotoolsPackage` passes `['autoconf','automake','libtool']`

For convenience, it's spec argument is also extended to take a list of
specs, which allows many packages which always require `autoreconf` to
add this to their `package.py` instead:

    `depends_on(AutotoolsPackage.autoreconf_depends, type='build')`

mwkrentel commented:
@bernhardkaindl  Yes, many autotools packages behave this way, but certainly not all.
For hpctoolkit, we include the generated files in all commits, not
just numbered releases, and we've been doing this long before spack
ever existed.

So, it's Ok for this to be the default, but I'd like there to be a
1-line knob to skip adding these dependencies for the packages that
don't use them and in fact, ignore them.

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -413,24 +413,38 @@ def _execute_conflicts(pkg):
+                return
+
+        if isinstance(specs, list):
+            for spec in specs:
+                _depends_on(pkg, spec, when=when, type=type, patches=patches)
```
I wouldn't accept a list of arguments, but rather do an explicit cycle where the dependency is declared. The API now has more possible inconsistencies in the arguments space. For instance, are `patches` supposed to be applicable to all arguments?

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -413,24 +413,38 @@ def _execute_conflicts(pkg):
         type (str or tuple): str or tuple of legal Spack deptypes
         patches (typing.Callable or list): single result of ``patch()`` directive, a
             ``str`` to be passed to ``patch``, or a list of these
+        skip (list): packages for which this depends_on call shall do nothing. This
+            is helpful when a base class of other packages like AutotoolsPackage wishes
+            to define dependencies for all subclasses, except for these packages.
+            Example: Add autotools packages for develop builds, but not on themselfes.
```
We should probably think more about this `skip` argument. I am wondering if it makes sense to have an operation that is the opposite of `when` for instance:
```python
depends_on('foo', unless='platform=darwin')
```
Another thought is that here we are discarding metadata from the Python code, while we should probably codify it in ASP and let `clingo` decide whether this should be taken into account or not. This just to have a single place where constraints are resolved.

alalazo reviewed (changes requested):
A couple of thoughts on the proposed API change in the `depends_on` directive

haampie commented:
Fixing (1) seems like a good idea, but (2) is a bit controversial (packages may not have @master / @develop versions, and some that have them also have configure files already, also there is @main).

Can we split the PR and get the fix for (1) in? That would make backporting easier too.

haampie commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -258,7 +259,10 @@ def __init__(self, stream):
 
     def write(self, data):
-        self.stream.write(data)
+        try:
+            self.stream.write(data)
+        except UnicodeEncodeError:
+            self.stream.write(str(codecs.encode(data, 'utf-8')))
```
Was  this commited by accident?

bernhardkaindl replied:
Sorry, yes this was an accident, I'll remove it. I'll mark this PR as draft until everything is clear.

haampie commented on lib/spack/spack/build_systems/autotools.py:
```diff
@@ -259,7 +272,8 @@ def autoreconf(self, spec, prefix):
         # Else try to regenerate it
         autotools = ['m4', 'autoconf', 'automake', 'libtool']
-        missing = [x for x in autotools if x not in spec]
+        # Previously, spec had only dependencies. Now they are in spec._dependencies
+        missing = [x for x in autotools if x not in spec._dependencies]
```
You should use the public api and limit to direct build type dependencies: `spec.dependencies(deptype='build')` and then never use `in` as it seems to lead to false positives indeed:

```suggestion
        build_deps = [d.name for d in s.dependencies(deptype='build')]
        missing = [x for x in autotools if x not in build_deps]
```

haampie replied:
Would also be good to add a test for this...

bernhardkaindl replied:
Looks great, much better! With that, I get `NameError: name 's' is not defined`

I assume you mean `spec.dependencies`, like this:
```py
build_deps = [d.name for d in spec.dependencies(deptype='build')]
```
Another improvement proposal: 

As autoconf has m4 as run-dependency, m4 does not have to be added as `depends_on()` explicitly. `depends_on(autoconf)1 implies that m4 (indirectly) is provided as well, so packages do not have to add it explicitly to each package:
```diff
-        autotools = ['m4', 'autoconf', 'automake', 'libtool']
+        autotools = ['autoconf', 'automake', 'libtool']
```
This removes the need of having this in all master-building packages:
```py
+    depends_on('m4', when='@master', type='build')
```
Just this is enough (tested)
```py
+    depends_on('autoconf', when='@master', type='build')
+    depends_on('automake', when='@master', type='build')
+    depends_on('libtool',  when='@master', type='build')
```
I'll push that change to PR #26101 first, then continue here when only the check in #26101 is approved and fully tested in all the tests (and yes, we'll likely want/need a test for this check too). Any hints and help would be welcome.

bernhardkaindl committed:
build_systems/autotools.py: Fix detection of missing tools for autoreconf

If the `configure` script is not present, the `autoreconf()` phase runs
the `autoreconf` tool, generating the `configure` script and other files.

Because `autoconf`, `automake` and `libtool` are needed to run `autoreconf`,
`AutotoolsPackage.autoreconf()` attempts to check that these are available.

Issue #26005 shows that this check currently does not work properly:
It needs to look for the required dependencies in `specs._dependencies`
instead of currently in 'specs'.

Therefore it currently wrongly assumes that the autotools are added
as dependencies, when in reality, they are not. If the autotools are
then not installed properly by the host system, strange errors like
the error shown in issue #26005 occur. This fixes the dependency check.

bernhardkaindl committed:
RFC: autotools.py: add automatc depends on autotools when='@master,develop'

`AutotoolsPackage`, if the `configure` script is not present, runs the
autoreconf phase to to generate the `configure` script and other files.

For autoreconf to be guaranteed to work, depends are needed when
configure needs to be generated for when='@master'(from git clones):
```
    depends_on('autoconf', when='@master', type='build')
    depends_on('automake', when='@master', type='build')
    depends_on('libtool',  when='@master', type='build')
```

Adding this to each package.py currenty the only possiblity. Maybe it
is possible to add this automatically by `AutotoolsPackage` itself.

For m4, autoconf, automake and libtool, those need to be skipped
as that would create a circular dependency for those.

This commit adds an optional argument 'skip' to depends_on() which
AutotoolsPackage uses to pass `['m4','autoconf','automake','libtool']`,
but there is the proposal to call it 'unless=' and be more flexible.

For convenience, the depends_on(spec) is extended to allow for list
of specs, so many dependencies can be added on one line. Packages
which always have to run autoreconf, can then use this instead:

    depends_on(AutotoolsPackage.autoreconf_depends, type='build')

It is very likely this commit can be further improved, but it can serve
as a starting point for possible refinements.

bernhardkaindl committed:
autotools.py/autoreconf: Show the depends_on()s to add to the package

haampie commented:
Closing this, as the contribution here is mainly array support in depends_on, which is unlikely to be approved given @alalazo's comment, and also I'm a fan of declarative package.py files, and fear that supporting arrays will lead people to write imperative code to create an array followed by a single depends_on.
